toolchain/lexer/tokenized_buffer.cpp

```diff
@@ -410,28 +428,80 @@ class TokenizedBuffer::Lexer {
         return;
       }
 
-      open_groups_.pop_back();
-      CARBON_DIAGNOSTIC(
-          MismatchedClosing, Error,
-          "Closing symbol does not match most recent opening symbol.");
-      token_emitter_.Emit(opening_token, MismatchedClosing);
-
-      CARBON_CHECK(!buffer_.tokens().empty())
-          << "Must have a prior opening token!";
-      Token prev_token = buffer_.tokens().end()[-1];
-
-      // TODO: do a smarter backwards scan for where to put the closing
-      // token.
-      Token closing_token = buffer_.AddToken(
-          {.kind = opening_kind.GetClosingSymbol(),
-           .has_trailing_space = buffer_.HasTrailingWhitespace(prev_token),
-           .is_recovery = true,
-           .token_line = current_line_,
-           .column = current_column_});
-      TokenInfo& opening_token_info = buffer_.GetTokenInfo(opening_token);
-      TokenInfo& closing_token_info = buffer_.GetTokenInfo(closing_token);
-      opening_token_info.closing_token = closing_token;
-      closing_token_info.opening_token = opening_token;
+      LineInfo& opening_token_line_info =
+          buffer_.GetLineInfo(buffer_.GetTokenInfo(opening_token).token_line);
+
+      auto is_current_line_shallower_than_opening_token = [&]() {
```

Comment:
I think this would be a little easier to read as either a helper function `ShouldCloseGroup` or straight bools -- I think the lambdas probably won't be helping performance much, particularly since in erroneous code it's probably okay to have compilation be marginally slower. But note too you could write something like:

```
bool close_invalid = false;
if (opening_token_line_info.start == current_line_info_->start) {
  close_invalid = kind.IsClosingSymbol();
} else {
  close_invalid = opening_token_line_info.indent >= current_line_info_->indent;
}
```

or as a ternary:

```
if (kind == TokenKind::EndOfFile() ||
    (opening_token_line_info.start == current_line_info_->start) ?
      kind.IsClosingSymbol() :
      opening_token_line_info.indent >= current_line_info_->indent) {
```